crates/db_schema/replaceable_schema/utils.sql

```diff
@@ -0,0 +1,146 @@
+--   * `select_old_rows`
+--   * `select_new_rows`
+--   * `select_old_and_new_rows` with 2 columns:
+--       1. `count_diff`: `-1` for old rows and `1` for new rows, which can be used with `sum` to get the number
+--          to add to a count
+--       2. (same name as the trigger's table): the old or new row as a composite value
+CREATE PROCEDURE r.create_triggers (table_name text, function_body text)
+LANGUAGE plpgsql
+AS $a$
+DECLARE
+    defs text := $$
+    -- Delete
+    CREATE FUNCTION r.thing_delete_statement ()
+        RETURNS TRIGGER
+        LANGUAGE plpgsql
+        AS function_body_delete;
+    CREATE TRIGGER delete_statement
+        AFTER DELETE ON thing REFERENCING OLD TABLE AS select_old_rows
+        FOR EACH STATEMENT
+        EXECUTE FUNCTION r.thing_delete_statement ( );
+    -- Insert
+    CREATE FUNCTION r.thing_insert_statement ( )
+        RETURNS TRIGGER
+        LANGUAGE plpgsql
+        AS function_body_insert;
+    CREATE TRIGGER insert_statement
+        AFTER INSERT ON thing REFERENCING NEW TABLE AS select_new_rows
+        FOR EACH STATEMENT
+        EXECUTE FUNCTION r.thing_insert_statement ( );
+    -- Update
+    CREATE FUNCTION r.thing_update_statement ( )
+        RETURNS TRIGGER
+        LANGUAGE plpgsql
+        AS function_body_update;
+    CREATE TRIGGER update_statement
+        AFTER UPDATE ON thing REFERENCING OLD TABLE AS select_old_rows NEW TABLE AS select_new_rows
+        FOR EACH STATEMENT
+        EXECUTE FUNCTION r.thing_update_statement ( );
+    $$;
+    select_old_and_new_rows text := $$ (
+        SELECT
+            -1 AS count_diff,
+            old_table::thing AS thing
+        FROM
+            select_old_rows AS old_table
+        UNION ALL
+        SELECT
+            1 AS count_diff,
+            new_table::thing AS thing
+        FROM
+            select_new_rows AS new_table) $$;
+    empty_select_new_rows text := $$ (
+        SELECT
+            *
+        FROM
+            -- Real transition table
+            select_old_rows
+        WHERE
+            FALSE) $$;
+    empty_select_old_rows text := $$ (
+        SELECT
+            *
+        FROM
+            -- Real transition table
+            select_new_rows
+        WHERE
+            FALSE) $$;
+    BEGIN
+        function_body := replace(function_body, 'select_old_and_new_rows', select_old_and_new_rows);
+        -- `select_old_rows` and `select_new_rows` are made available as empty tables if they don't already exist
+        defs := replace(defs, 'function_body_delete', quote_literal(replace(function_body, 'select_new_rows', empty_select_new_rows)));
+        defs := replace(defs, 'function_body_insert', quote_literal(replace(function_body, 'select_old_rows', empty_select_old_rows)));
+        defs := replace(defs, 'function_body_update', quote_literal(function_body));
+        defs := replace(defs, 'thing', table_name);
+        EXECUTE defs;
+END;
+$a$;
```

Comment:
this seems reasonable though i have a hard time predicting whether these new (statement-level) triggers will have the same performance as the old row-level triggers, since in production all of these are executed for only single rows and if there's overhead due to the changes it might have significant impact

remember that activity receiving for example is sequential so every synchronous delay delays all of federation receiving

Reply:
For single-row statements, the only difference in the new triggers is the way it determines that the count changes are -1, 0, or 1. This process should be almost nothing compared to the update that is done afterwards, which is unchanged. So I don't expect a measurable difference in performance.